fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java

```diff
@@ -0,0 +1,123 @@
+import net.minecraft.util.Pair;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+public final class FluidFlowEvents {
+	private FluidFlowEvents() {
+	}
+
+	private static final Map<Block, Map<Block, List<Pair<Direction, Event<FluidFlowInteractionEvent>>>>> EVENT_MAP = new HashMap<>();
+
+	/**
+	 * Registers a new event on a fluid flow. The same two blocks can register a different event for different directions, but the same directions will run the event that was registered first.
+	 *
+	 * @param flowingBlock          The fluid block that flowed.
+	 * @param interactionBlock      The block in one of the {@code interactionDirections}.
+	 * @param interactionDirections The direction to search for {@code interactionBlock}.
+	 * @param interactionEvent      The event to run when the conditions are met.
+	 */
+	public static void register(Block flowingBlock, Block interactionBlock, Direction[] interactionDirections, FluidFlowInteractionEvent interactionEvent) {
+		Map<Block, List<Pair<Direction, Event<FluidFlowInteractionEvent>>>> flowBlockEvents = EVENT_MAP.getOrDefault(flowingBlock, new HashMap<>());
+		List<Pair<Direction, Event<FluidFlowInteractionEvent>>> interactionEvents = flowBlockEvents.getOrDefault(interactionBlock, new ArrayList<>());
+
+		if (interactionEvents.isEmpty()) {
+			for (Direction direction : Direction.values()) {
+				interactionEvents.add(new Pair<>(direction, EventFactory.createArrayBacked(FluidFlowInteractionEvent.class, fluidFlowInteractionEvents -> (flowingBlockState, interactingBlockState, flowPos, world) -> {
+					for (FluidFlowInteractionEvent event : fluidFlowInteractionEvents) {
+						if (!event.onFlow(flowingBlockState, interactingBlockState, flowPos, world)) {
+							return false;
+						}
+					}
+
+					return true;
+				})));
+			}
+		}
+
+		for (Pair<Direction, Event<FluidFlowInteractionEvent>> pair : interactionEvents) {
+			for (Direction direction : interactionDirections) {
+				if (pair.getLeft() == direction) {
+					pair.getRight().register(interactionEvent);
+				}
+			}
+		}
+
+		if (!flowBlockEvents.containsKey(interactionBlock)) {
+			flowBlockEvents.put(interactionBlock, interactionEvents);
+		}
+
+		if (!EVENT_MAP.containsKey(flowingBlock)) {
+			EVENT_MAP.put(flowingBlock, flowBlockEvents);
+		}
+	}
+
+	/**
+	 * Gets the event from the following blocks and direction.
+	 *
+	 * @param flowingBlock         The fluid block that flowed.
+	 * @param interactionBlock     The block it interacts with.
+	 * @param interactionDirection The interaction direction
+	 * @return An event if the conditions are met, otherwise {@code null}
+	 */
+	public static @Nullable Event<FluidFlowInteractionEvent> getEvent(Block flowingBlock, Block interactionBlock, Direction interactionDirection) {
+		if (EVENT_MAP.containsKey(flowingBlock)) {
```

Comment:
just check if the get result is null. extra contains check offers no real benefits